Add witai parser

FabioRosado opened the pull request:
Still haven't attempted to create tests for it.

edit: The bad auth code that I was getting was because I placed the access-token key between  "", the access token needs to be inserted without them.
 
edit2: match_witai now works if imported in a skill

jacobtomlinson commented:
Nice! Have a look at #263 and #267 which add LUIS and Lex support respectively. You should see what they've done and hopefully anything you've missed.

With the matcher issue is that because you named the decorator `match_witai_intent`?

FabioRosado commented:
Should I name `match_witai_intent` differently?

jacobtomlinson commented:
It's up to you really. The only reason the apiai one is called `match_apiai_intent` is because there is also `match_apiai_action`, if that didn't exist I would probably have just called it `match_apiai`.

I'm happy either way, just ensure you are consistent and that it is documented.

FabioRosado commented on opsdroid/parsers/witai.py:
```diff
@@ -0,0 +1,68 @@
+                            _LOGGER.exception("Exception when parsing '" +
+                                              message.text +
+                                              "' against skill '" +
+                                              result + "'")
```
I'm writing the tests and for this bit to work I have to do one of the following:
`result['entities']['intent'][0]['value']` which makes the line too long or simply do `str(result)` and pass the whole json received, which one would you rather use?

jacobtomlinson replied:
I would say the first probably.

FabioRosado commented:
I changed the name of the matcher as suggested, I've also added the tests for wit.ai based on api.ai ones. I've added the documentation for wit.ai but I want to play around with `message.witai` and check how to use a skill other information returned by witai.

I've also tested the wit.ai matcher/parser and it seems to work nicely with the weather skill so I guess it should work fine with other skills as well.

I've finally understood why one of the docstring tests where failing, sorry for the testing spam. 
Let me know what you think and if I need to change anything.

jacobtomlinson commented:
This is really good work!

Here's a few points to get the last couple of steps so it can be merged:

- Travis is failing because the linter is complaining about a broad exception. Check out the lint exception comments in the [other parsers](https://github.com/opsdroid/opsdroid/blob/master/opsdroid/parsers/luisai.py#L32).
- Can you throw in a test for a response with no entities to cover [43-45](https://coveralls.io/builds/13739572/source?filename=opsdroid%2Fparsers%2Fwitai.py#L43) in `witai.py`. The other missed lines are fine, they are missing in the other parsers too.
- Can you update the example in the docs to use an entity like a location from the response. For example you could ask wit.ai "what's the weather in London" and then get the location entity `London` and then use it in your example skill.

FabioRosado commented:
I've followed your suggestions and updated the commit.
Let me know if the docs make sense or if they are too confusing. I'm going to start reading about tests to try and figure out how to test for `ClientOSError`

jacobtomlinson commented:
Great I'll take a look at this tomorrow!

Why don't you have a look at those tests under a new PR. I'm happy to merge this one once I've tested it locally.

jacobtomlinson commented on opsdroid/parsers/witai.py:
```diff
@@ -0,0 +1,76 @@
+                          "forget to create one?")
+            return
+
+        confidence = result['entities']['intent'][0]['confidence']
```
If the string doesn't match an intent then there will be no intent paramater in the response resulting in an error. This needs to be caught.

```python
Traceback (most recent call last):
  File "/opt/boxen/homebrew/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/Users/jacob/Projects/opsdroid/opsdroid/opsdroid/parsers/witai.py", line 51, in parse_witai
    confidence = result['entities']['intent'][0]['confidence']
KeyError: 'intent'
```

FabioRosado commented:
I've added a try/except on that piece of code that you mentioned @jacobtomlinson, I've also added a test for it. At the moment I'm setting confidence to 0.0 because if there is no intent then probably there  is no action to follow up.  Please let me know if this is okay or if you rather have it done in a different way.

I've also added the `test_parse_witai_clientOSError` to see if I could pass the ClientOSError, the coverage seem to increase 2%, does this means that this test worked? If so I'll update the other parsers to include it 👍

jacobtomlinson commented:
Yeah I like that! Annoyingly the linter is now complaining about some whitespace issues. Can you quickly tweak those.

FabioRosado commented:
Sorry for that I was giving it a go at the ClientOSError issue and messed up the spaces.
This attempt should pass the tests, I'll keep trying to play around with the ClientOSError to see if I can crack the issue 👍

jacobtomlinson commented:
Awesome good work!
